web/src/message_events.js

```diff
@@ -35,6 +37,129 @@ import * as unread_ops from "./unread_ops";
+                force_rerender: true,
+            });
+            continue;
+        }
+
+        // We need the message objects to determine if they match the filter.
+        const messages_to_fetch = [];
+        const messages = [];
+        for (const message_id of message_ids) {
+            const message = message_store.get(message_id);
+            if (message !== undefined) {
+                messages.push(message);
+            } else {
+                const first_id = msg_list.first().id;
+                const last_id = msg_list.last().id;
+                const has_found_newest = msg_list.data.fetch_status.has_found_newest();
+                const has_found_oldest = msg_list.data.fetch_status.has_found_oldest();
+
+                if (message_id > first_id && message_id < last_id) {
+                    // Need to insert message middle of the list.
+                    messages_to_fetch.push(message_id);
+                } else if (message_id < first_id && has_found_oldest) {
+                    // Need to insert message at the start of list.
+                    messages_to_fetch.push(message_id);
+                } else if (message_id > last_id && has_found_newest) {
+                    // Need to insert message at the end of list.
+                    messages_to_fetch.push(message_id);
+                }
+            }
+        }
+
+        // In most cases, we will only have one message to fetch which
+        // can happen without rerendering the view.
+        // In case of multiple messages, we just rerender the view
+        // since it is likely to use similar amount of resources to
+        // fetching the messages and rerendering the view.
+        if (messages_to_fetch > 1) {
+            const current_selected_id = msg_list.selected_id();
+            message_view.show(filter.terms(), {
+                then_select_id: current_selected_id,
+                trigger: "message property update",
+                force_rerender: true,
+            });
+        } else if (messages_to_fetch.length === 1) {
+            // Fetch the message and update the view.
+            channel.get({
+                url: "/json/messages/" + messages_to_fetch[0],
+                success(data) {
+                    message_helper.process_new_message(data.message);
+                    update_views_filtered_on_message_property(message_ids, term_type);
+                },
+            });
+        } else {
+            // We have all the messages locally, so we can update the view.
+            //
+            // Special case: For starred messages view, we don't remove
+            // messages that are no longer starred to avoid
+            // implementing an undo mechanism for that view.
+            if (
+                term_type === "is-starred" &&
+                _.isEqual(filter.sorted_term_types(), ["is-starred"])
+            ) {
+                msg_list.add_messages(messages);
+                continue;
+            }
```

Comment:
Maybe worth a TODO comment that a cleaner way to implement this might be to track which things have been unstarred in the starred messages view in this visit to the view, and have those stay? The old issue by Steve talks about this corner case.

Reply:
Added